### PR TITLE
AMLII-1105 Fix race condition during demultiplexer stop

### PR DESCRIPTION
**Pull request by vickenty**
### What does this PR do?
Synchronize access to senders

AgentDemultiplexer will reset the senders field to zero on shutdown.

Collector workers may try to access it during or after demuliplexer
shutdown and will experience data races or crashes on nil
field. Collector deliberately doesn't hang up on finalizing
workers (there is a 2 second timeout).

### Motivation

Fix a flaky test

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/aggregator/demultiplexer_agent.go [hunk 1]
```diff
@@ -607,3 +612,55 @@ func (d *AgentDemultiplexer) Aggregator() *BufferedAggregator {
...
+	d.senders.DestroySender(id)
+}
+
+// GetDefaultSender returns a default sender.
+func (d *AgentDemultiplexer) GetDefaultSender() (sender.Sender, error) {
+	d.m.Lock()
+	defer d.m.Unlock()
+
+	if d.senders == nil {
+		return nil, errors.New("demultiplexer is stoped")
```
Comment (remeh) [inline pkg/aggregator/demultiplexer_agent.go:662]: typo in stoped (there is the same in other functions and in `test_common.go`)

**Review by remeh — COMMENTED** [review]
I noticed that the senders object already have some locking: https://github.com/DataDog/datadog-agent/blob/main/pkg/aggregator/sender.go#L381

Did you noticed race conditions on this `senders` map? https://github.com/DataDog/datadog-agent/blob/main/pkg/aggregator/sender.go#L92

**Comment by vickenty** [comment]
> I noticed that the senders object already have some locking: https://github.com/DataDog/datadog-agent/blob/main/pkg/aggregator/sender.go#L381
> 
> Did you noticed race conditions on this `senders` map? https://github.com/DataDog/datadog-agent/blob/main/pkg/aggregator/sender.go#L92

No, the race is on the `senders` field in the AgentDemultiplexer. It is set to nil in `Stop()` [here](https://github.com/datadog/datadog-agent/blob/52480da68d5e96a1595e31378ce33c0d51281b70/pkg/aggregator/demultiplexer_agent.go#L437), while the collector workers may still access it via the global sender getters:

```
WARNING: DATA RACE
Read at 0x00c00032f688 by goroutine 251:
  github.com/DataDog/datadog-agent/pkg/aggregator.(*TestAgentDemultiplexer).GetDefaultSender()
      <autogenerated>:1 +0x64
[...]
Previous write at 0x00c00032f688 by goroutine 245:
  github.com/DataDog/datadog-agent/pkg/aggregator.(*AgentDemultiplexer).Stop()
      /Users/runner/go/src/github.com/DataDog/datadog-agent/pkg/aggregator/demultiplexer_agent.go:432 +0x753
```

**Review by remeh — APPROVED** [review]
Thanks for the fix 🙇

**Review by carlosroman — APPROVED** [review]

**Pull request merged**